src/rpdk/guard_rail/rule_library/core/schema-linter-core-rules.guard

```diff
@@ -27,67 +20,132 @@ rule ensure_primary_identifier_exists_and_not_empty
 
 rule ensure_primary_identifier_is_read_or_create_only when ensure_primary_identifier_exists_and_not_empty
 {
-    primaryIdentifier[*] {
-        this IN %read OR
-        this IN %create
+    when readOnlyProperties exists {
+        let read = readOnlyProperties
+        when createOnlyProperties exists {
+            let create = createOnlyProperties
+            primaryIdentifier[*] {
+                this IN %read OR
+                this IN %create
+                <<
+                {
+                    "result": "NON_COMPLIANT",
+                    "check_id": "PID003",
+                    "message": "primaryIdentifier MUST be either readOnly or createOnly"
+                }
+                >>
+            }
+        }
+
+        when createOnlyProperties !exists {
+            primaryIdentifier[*] {
+                this IN %read
+                <<
+                {
+                    "result": "NON_COMPLIANT",
+                    "check_id": "PID003",
+                    "message": "primaryIdentifier MUST be either readOnly or createOnly"
+                }
+                >>
+            }
+        }
+    }
+
+    when readOnlyProperties !exists {
+        createOnlyProperties !exists
```

Comment:
Not familiar with `.guard` syntax either, but just in case. Is it not necessary to use the `when` clause here?

Reply:
yea it's basically skipped when it does not hold. In this case if `createOnlyProperties` and `readOnlyProperties` are not declared then primary identifier cant be part of them which is against our best practices